spring-rabbit/src/main/java/org/springframework/amqp/rabbit/support/ClosingRecoveryListener.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.amqp.rabbit.support;
+
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.amqp.rabbit.connection.ChannelProxy;
+
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Recoverable;
+import com.rabbitmq.client.RecoveryListener;
+import com.rabbitmq.client.impl.recovery.AutorecoveringChannel;
+
+/**
+ * A {@link RecoveryListener} that closes the recovered channel, to avoid
+ * orphaned consumers.
+ *
+ * @author Gary Russell
+ * @since 1.7.10
+ *
+ */
+public final class ClosingRecoveryListener implements RecoveryListener {
+
+	private static final Log logger = LogFactory.getLog(ClosingRecoveryListener.class);
+
+	private static final RecoveryListener INSTANCE = new ClosingRecoveryListener();
+
+	private static final ConcurrentMap<AutorecoveringChannel, Boolean> hasListener = new ConcurrentHashMap<>();
```

Comment:
Looks like you want to have a `Collections.newSetFromMap()`.

Reply:
No; I originally had `ConcurrentHashMap.newKeySet()`, but I need `putIfAbsent()`.

Reply:
Well, but our `boolean` value is idempotent there. Only what bother us is exactly if there was value before or not.
See `HashSet` impl:
```
  * @return <tt>true</tt> if this set did not already contain the specified
  * element
  */
  public boolean add(E e) {
        return map.put(e, PRESENT)==null;
  }
```

For me this is exactly what you do with:
```
if (hasListener.putIfAbsent(autorecoveringChannel, Boolean.TRUE) == null) {
```

So, essentially you implement here some small `Set` logic, but let it be if you insist. At least we get some performance gain skipping `Set` operation and moving to `Map` one directly. 😉

Reply:
Yes; that was my thought, and `newKeySet()` does exactly that anyway (back the set with a map with boolean values, always true).